src/NLog/Targets/FileArchiveStyleFactory.cs

```diff
@@ -0,0 +1,758 @@
+        }
+
+        public List<DateAndSequenceArchive> GetExistingArchiveFiles(string archiveFilePath)
+        {
+            string archiveFolderPath = Path.GetDirectoryName(archiveFilePath);
+            FileNameTemplate archiveFileNameTemplate = GenerateFileNameTemplate(archiveFilePath);
+            string archiveFileMask = GenerateFileNameMask(archiveFilePath, archiveFileNameTemplate);
+
+            var existingArchiveFiles = new List<DateAndSequenceArchive>();
+            if (string.IsNullOrEmpty(archiveFileMask))
+                return existingArchiveFiles;
+
+            DirectoryInfo directoryInfo = new DirectoryInfo(archiveFolderPath);
+            if (!directoryInfo.Exists)
+                return existingArchiveFiles;
+
+#if SILVERLIGHT && !WINDOWS_PHONE
+            var existingFiles = directoryInfo.EnumerateFiles(archiveFileMask);
+#else
+            var existingFiles = directoryInfo.GetFiles(archiveFileMask);
+#endif
+            foreach (var fileInfo in existingFiles)
+            {
+                var archiveFileInfo = GenerateArchiveFileInfo(fileInfo, archiveFileNameTemplate);
+                if (archiveFileInfo != null)
+                    existingArchiveFiles.Add(archiveFileInfo);
+            }
+
+            if (existingArchiveFiles.Count > 1)
+                existingArchiveFiles.Sort((x, y) => (x.Date != y.Date && !x.HasSameFormattedDate(y.Date)) ? x.Date.CompareTo(y.Date) : x.Sequence.CompareTo(y.Sequence) != 0 ? x.Sequence.CompareTo(y.Sequence) : string.CompareOrdinal(x.FileName, y.FileName));
+            return existingArchiveFiles;
+        }
+
+        protected virtual FileNameTemplate GenerateFileNameTemplate(string archiveFilePath)
+        {
+            return new FileNameTemplate(Path.GetFileName(archiveFilePath));
+        }
+        protected virtual string GenerateFileNameMask(string archiveFilePath, FileNameTemplate fileTemplate)
+        {
+            if (fileTemplate != null)
+                return fileTemplate.ReplacePattern("*");
+            else
+                return string.Empty;
+        }
+        protected abstract DateAndSequenceArchive GenerateArchiveFileInfo(FileInfo archiveFile, FileNameTemplate fileTemplate);
+        public abstract DateAndSequenceArchive GenerateArchiveFileName(string archiveFilePath, DateTime archiveDate, List<DateAndSequenceArchive> existingArchiveFiles);
+
+        public virtual IEnumerable<DateAndSequenceArchive> CheckArchiveCleanup(string archiveFilePath, List<DateAndSequenceArchive> existingArchiveFiles, int maxArchiveFiles)
+        {
+            if (maxArchiveFiles <= 0)
+                yield break;
+
+            _lastArchiveFilePath = archiveFilePath; // Cache that we have just performed cleanup for this archive-path
+
+            if (existingArchiveFiles.Count == 0 || existingArchiveFiles.Count < maxArchiveFiles)
+                yield break;
+
+            for (int i = 0; i < existingArchiveFiles.Count - maxArchiveFiles; ++i)
```

Comment:
`i++` instead of `++i` ?

I'm curious, what's the benefit of `++i` in a for loop?

edit: it was already there I think